torch/optim/_functional.py

```diff
@@ -357,3 +357,48 @@ def asgd(params: List[Tensor],
             ax.add_(param.sub(ax).mul(mu))
         else:
             ax.copy_(param)
+
+
+def nadam(params: List[Tensor],
+          grads: List[Tensor],
+          exp_avgs: List[Tensor],
+          exp_avg_sqs: List[Tensor],
+          m_schedules: List[float],
+          state_steps: List[int],
+          *,
+          beta1: float,
+          beta2: float,
+          lr: float,
+          weight_decay: float,
+          schedule_decay: float,
+          eps: float):
+    r"""Functional API that performs NAdam algorithm computation.
+
+    See :class:`~torch.optim.NAdam` for details.
+    """
+
+    for i, param in enumerate(params):
+        grad = grads[i]
+        exp_avg = exp_avgs[i]
+        exp_avg_sq = exp_avg_sqs[i]
+        m_schedule = m_schedules[i]
+        step = state_steps[i]
+
+        bias_correction2 = 1 - beta2 ** step
+
+        if weight_decay != 0:
+            grad = grad.add(param, alpha=weight_decay)
+
+        # calculate the momentum cache \mu^{t} and \mu^{t+1}
+        momentum_cache_t = beta1 * (1. - 0.5 * (0.96 ** (step * schedule_decay)))
+        momentum_cache_t_1 = beta1 * (1. - 0.5 * (0.96 ** ((step + 1) * schedule_decay)))
```

Comment:
Where are the 0.96 coming from? I do not see it in the [paper](https://openreview.net/pdf?id=OM0jvwB8jIp57ZJjtNEZ).

Reply:
Yes, I agree that there is no clear way of describing momentum_cache computation in the paper, and we can see mu_1, mu_2, ... mu_T left as hyperparameter. However, i have seen many times in practice this is the common choice, and even Tensorflow/Keras follows the same practise :  https://github.com/tensorflow/tensorflow/blob/f9d386849581d15d72f6f1f96f12aac230a8edbe/tensorflow/python/keras/optimizer_v2/nadam.py#L113

Reply:
The number is actually taken from the owner of the algorithm - Timothy Dozat's original report : http://cs229.stanford.edu/proj2015/054_report.pdf  . My understanding Keras and many other libraries in practice taking the number from this report.